### PR TITLE
Basic markdown formatting for comments

### PR DESCRIPTION
**Pull request by flovilmart**
Hit me up with all your might about code style, Just put that quickly together as POC.

Probable improvements:
- [x] insert at beginning of lines for H's etc...
- [x] move cursor inside for code blocks, blocks + text formatting

### CODE REVIEW

File: Classes/NewComment/MDownFormattingController.swift [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import Foundation
+import UIKit
+
+struct MDownFormattingElement {
+    let name: String
+    let replacements: (String?, String?) // before / after
```
Comment (BasThomas) [inline Classes/NewComment/MDownFormattingController.swift]: In which case is the `before` nil?

File: Classes/NewComment/MDownFormattingController.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+
+let baseElements = [
+    MDownFormattingElement(name: "Code Block", replacements: ("```\n", "\n```")),
```
Comment (BasThomas) [inline Classes/NewComment/MDownFormattingController.swift]: Might want to localize the names?

Comment (flovilmart) [inline Classes/NewComment/MDownFormattingController.swift]: Yep 👍

File: Classes/NewComment/MDownFormattingController.swift (hunk 1 above)
Comment (flovilmart) [inline Classes/NewComment/MDownFormattingController.swift]: Uhmmm for now, none, I can probably remove it.

File: Classes/NewComment/NewCommentViewController.swift [hunk 3]
```diff
@@ -68,6 +69,11 @@ final class NewCommentViewController: UIViewController, UITextViewDelegate, AddC
         textView.delegate = self
         textView.font = Styles.Fonts.body
         textView.backgroundColor = .clear
+        textView.inputAccessoryView = formattingController.view
+        formattingController.textView = textView
+        // need to constraint here, maybe better idea swhere else
+        // Otherwise takes all screen
+        textView.inputAccessoryView?.frame = CGRect(origin: .zero, size: CGSize(width: 0.0, height: 44.0))
```
Comment (flovilmart) [inline Classes/NewComment/NewCommentViewController.swift]: I would prefer if that was handled but the controller, but I don't seem to get the constraint working :/

**Comment by flovilmart** [comment]
For the insertion at beginning of line, not sure if I'm gonna do it now, as it required parsing the text, find the previous "\n" etc...

**Comment by flovilmart** [comment]
Let me know if there's anything else you were thinking about.

File: Classes/NewComment/CommentFormatterAccessoryView.swift [hunk 4]
```diff
@@ -0,0 +1,164 @@
+import Foundation
```
Comment (rnystrom) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: missing header comment

File: Classes/NewComment/CommentFormatterAccessoryView.swift [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    private func setupInitialConstraintsIfNeeded() {
+        if hasInitialConstraints { return }
+        hasInitialConstraints = true
+        translatesAutoresizingMaskIntoConstraints = false
+
+        scrollView.translatesAutoresizingMaskIntoConstraints = false
+        scrollView.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
+        scrollView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
+        scrollView.topAnchor.constraint(equalTo: topAnchor).isActive = true
+        scrollView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
```
Comment (rnystrom) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: Could we use SnapKit to keep things consistent w/ the rest of the codebase?

File: Classes/NewComment/CommentFormatterAccessoryView.swift [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+        stackView.translatesAutoresizingMaskIntoConstraints = false
+        stackView.leadingAnchor.constraint(equalTo: scrollView.leadingAnchor, constant: 12).isActive = true
+        stackView.trailingAnchor.constraint(equalTo: scrollView.trailingAnchor, constant: -12).isActive = true
+        stackView.topAnchor.constraint(equalTo: scrollView.topAnchor, constant: 6).isActive = true
+        stackView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor).isActive = true
```
Comment (rnystrom) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: same nit re SnapKit

File: Classes/NewComment/CommentFormatterAccessoryView.swift [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        textView.replace(insertionRange, withText: replacementText)
+        if range.start == range.end, // single cursor (no selection)
+            let position = textView.position(from: range.start, // advance by the inserted before
+                                             offset: before.lengthOfBytes(using: .utf8)) {
+            textView.selectedTextRange = textView.textRange(from: position, to: position) // single cursor
+        }
+    }
+
+    deinit {
+        print("DEINIT!!")
```
Comment (rnystrom) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: lol, remove?

File: Classes/NewComment/CommentFormatterAccessoryView.swift [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+]
+
+private func skinnedButton(_ button: UIButton) -> UIButton {
+    button.contentEdgeInsets = UIEdgeInsets(top: 5.0, left: 20.0, bottom: 5.0, right: 20.0)
+    button.backgroundColor = Styles.Colors.Blue.medium.color
+    button.layer.cornerRadius = 16.0 // just so we don't need recompute
+    button.layer.masksToBounds = true
+    button.setTitleColor(.white, for: .normal)
+    return button
+}
```
Comment (rnystrom) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: I'll probably tweak these settings a little bit, not crazy about the blue background. Going to restyle to be more like the keyboard buttons

File: Classes/NewComment/CommentFormatterAccessoryView.swift [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    MDownFormattingElement(name: "H4", replacements: ("#### ", nil), atLineStart: true),
+    MDownFormattingElement(name: "mdown.bold", replacements: ("**", "**")),
+    MDownFormattingElement(name: "mdown.emphasis", replacements: ("*", "*")),
+    MDownFormattingElement(name: "mdown.strike", replacements: ("~~", "~~"))
+]
+
+private func skinnedButton(_ button: UIButton) -> UIButton {
+    button.contentEdgeInsets = UIEdgeInsets(top: 5.0, left: 20.0, bottom: 5.0, right: 20.0)
+    button.backgroundColor = Styles.Colors.Blue.medium.color
+    button.layer.cornerRadius = 16.0 // just so we don't need recompute
```
Comment (rnystrom) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: Can we use values consistent with `Styles.Sizes`?

**Comment by flovilmart** [comment]
Was thinking about usernames as well, went stackview instead of collection. But collection may be better for usernames. Also will need backing for more datasource OR we could have multiple accessories based on keyboard input in delegate. I’ll add the nits, let me know if you want me to factor the code differently :)

File: Classes/NewComment/CommentFormatterAccessoryView.swift (hunk 9 above)
Comment (flovilmart) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: Yes!

File: Classes/NewComment/CommentFormatterAccessoryView.swift (hunk 4 above)
Comment (flovilmart) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: I'll add

File: Classes/NewComment/CommentFormatterAccessoryView.swift (hunk 5 above)
Comment (flovilmart) [inline Classes/NewComment/CommentFormatterAccessoryView.swift]: makes sense, never used snap kit :)

**Comment by flovilmart** [comment]
@rnystrom do you want me to refactor as a UICollectionViewController / UICollectionView with an abstraction over the datasource / actions so it support easy insertion of any kind of text for future proofing?

**Comment by rnystrom** [comment]
@flovilmart that sounds awesome!

**Comment by flovilmart** [comment]
Alright let's do that then!
@rnystrom I see you'Ve implemented the name auto-complete which is awesome, should I still aim for a keyboard accessory with the same layout, (horizontal scroll / buttons) as the name autocomplete won't probably render in this keyboard accessory (if you intent to keep using the one you have already). In this case, I'm not sure why a refactoring / future-proofing for usernames if we don't plan to use the horizontal scrolling for other purposes. 

lmk

**Comment by rnystrom** [comment]
Ya we don't need to save this space for autocomplete anymore now that I'm using Slacks implementation!

**Comment by rnystrom** [comment]
Took this code and got it working!

**Pull request closed without merging**